File: spine-ts/core/src/Skin.ts
/******************************************************************************
 * Spine Runtimes License Agreement
 * Last updated January 1, 2020. Replaces all prior versions.
 *
 * Copyright (c) 2013-2020, Esoteric Software LLC
 *
 * Integration of the Spine Runtimes into software or otherwise creating
 * derivative works of the Spine Runtimes is permitted under the terms and
 * conditions of Section 2 of the Spine Editor License Agreement:
 * http://esotericsoftware.com/spine-editor-license
 *
 * Otherwise, it is permitted to integrate the Spine Runtimes into software
 * or otherwise create derivative works of the Spine Runtimes (collectively,
 * "Products"), provided that each user of the Products must obtain their own
 * Spine Editor license and redistribution of the Products in any form must
 * include this license and copyright notice.
 *
 * THE SPINE RUNTIMES ARE PROVIDED BY ESOTERIC SOFTWARE LLC "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL ESOTERIC SOFTWARE LLC BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES,
 * BUSINESS INTERRUPTION, OR LOSS OF USE, DATA, OR PROFITS) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THE SPINE RUNTIMES, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

module spine {
	/** Stores an entry in the skin consisting of the slot index, name, and attachment **/
	export class SkinEntry {
		constructor(public slotIndex: number, public name: string, public attachment: Attachment) { }
	}

	/** Stores attachments by slot index and attachment name.
	 *
	 * See SkeletonData {@link SkeletonData#defaultSkin}, Skeleton {@link Skeleton#skin}, and
	 * [Runtime skins](http://esotericsoftware.com/spine-runtime-skins) in the Spine Runtimes Guide. */
	export class Skin {
		/** The skin's name, which is unique across all skins in the skeleton. */
		name: string;

		attachments = new Array<Map<Attachment>>();
		bones = Array<BoneData>();
		constraints = new Array<ConstraintData>();

		constructor (name: string) {
			if (name == null) throw new Error("name cannot be null.");
			this.name = name;
		}

		/** Adds an attachment to the skin for the specified slot index and name. */
		setAttachment (slotIndex: number, name: string, attachment: Attachment) {
			if (attachment == null) throw new Error("attachment cannot be null.");
			let attachments = this.attachments;
			if (slotIndex >= attachments.length) attachments.length = slotIndex + 1;
			if (!attachments[slotIndex]) attachments[slotIndex] = { };
			attachments[slotIndex][name] = attachment;
		}

		/** Adds all attachments, bones, and constraints from the specified skin to this skin. */
		addSkin (skin: Skin) {
			for(let i = 0; i < skin.bones.length; i++) {
				let bone = skin.bones[i];
				let contained = false;
				for (let j = 0; j < this.bones.length; j++) {
					if (this.bones[j] == bone) {
						contained = true;
						break;
					}
				}
				if (!contained) this.bones.push(bone);
			}

			for(let i = 0; i < skin.constraints.length; i++) {
				let constraint = skin.constraints[i];
				let contained = false;
				for (let j = 0; j < this.constraints.length; j++) {
					if (this.constraints[j] == constraint) {
						contained = true;
						break;
					}
				}
				if (!contained) this.constraints.push(constraint);
			}

			let attachments = skin.getAttachments();
			for (let i = 0; i < attachments.length; i++) {
				var attachment = attachments[i];
				this.setAttachment(attachment.slotIndex, attachment.name, attachment.attachment);
			}
		}

		/** Adds all bones and constraints and copies of all attachments from the specified skin to this skin. Mesh attachments are not
		 * copied, instead a new linked mesh is created. The attachment copies can be modified without affecting the originals. */
		copySkin (skin: Skin) {
			for(let i = 0; i < skin.bones.length; i++) {
				let bone = skin.bones[i];
				let contained = false;
				for (let j = 0; j < this.bones.length; j++) {
					if (this.bones[j] == bone) {
						contained = true;
						break;
					}
				}
				if (!contained) this.bones.push(bone);
			}

			for(let i = 0; i < skin.constraints.length; i++) {
				let constraint = skin.constraints[i];
				let contained = false;
				for (let j = 0; j < this.constraints.length; j++) {
					if (this.constraints[j] == constraint) {
						contained = true;
						break;
					}
				}
				if (!contained) this.constraints.push(constraint);
			}

			let attachments = skin.getAttachments();
			for (let i = 0; i < attachments.length; i++) {
				var attachment = attachments[i];
				if (attachment.attachment == null) continue;
				if (attachment.attachment instanceof MeshAttachment) {
					attachment.attachment = attachment.attachment.newLinkedMesh();
					this.setAttachment(attachment.slotIndex, attachment.name, attachment.attachment);
				} else {
					attachment.attachment = attachment.attachment.copy();
					this.setAttachment(attachment.slotIndex, attachment.name, attachment.attachment);
				}
			}
		}

		/** Returns the attachment for the specified slot index and name, or null. */
		getAttachment (slotIndex: number, name: string): Attachment {
			let dictionary = this.attachments[slotIndex];
			return dictionary ? dictionary[name] : null;
		}

		/** Removes the attachment in the skin for the specified slot index and name, if any. */
		removeAttachment (slotIndex: number, name: string) {
			let dictionary = this.attachments[slotIndex];
			if (dictionary) dictionary[name] = null;
		}

		/** Returns all attachments in this skin. */
		getAttachments (): Array<SkinEntry> {
			let entries = new Array<SkinEntry>();
			for (var i = 0; i < this.attachments.length; i++) {
				let slotAttachments = this.attachments[i];
				if (slotAttachments) {
					for (let name in slotAttachments) {
						let attachment = slotAttachments[name];
						if (attachment) entries.push(new SkinEntry(i, name, attachment));
					}
				}
			}
			return entries;
		}

		/** Returns all attachments in this skin for the specified slot index. */
		getAttachmentsForSlot (slotIndex: number, attachments: Array<SkinEntry>) {
			let slotAttachments = this.attachments[slotIndex];
			if (slotAttachments) {
				for (let name in slotAttachments) {
					let attachment = slotAttachments[name];
					if (attachment) attachments.push(new SkinEntry(slotIndex, name, attachment));
				}
			}
		}

		/** Clears all attachments, bones, and constraints. */
		clear () {
			this.attachments.length = 0;
			this.bones.length = 0;
			this.constraints.length = 0;
		}

		/** Attach each attachment in this skin if the corresponding attachment in the old skin is currently attached. */
		attachAll (skeleton: Skeleton, oldSkin: Skin) {
			let slotIndex = 0;
			for (let i = 0; i < skeleton.slots.length; i++) {
				let slot = skeleton.slots[i];
				let slotAttachment = slot.getAttachment();
				if (slotAttachment && slotIndex < oldSkin.attachments.length) {
					let dictionary = oldSkin.attachments[slotIndex];
					for (let key in dictionary) {
						let skinAttachment:Attachment = dictionary[key];
						if (slotAttachment == skinAttachment) {
							let attachment = this.getAttachment(slotIndex, key);
							if (attachment != null) slot.setAttachment(attachment);
							break;
						}
					}
				}
				slotIndex++;
			}
		}
	}
}

File: spine-libgdx/spine-libgdx/src/com/esotericsoftware/spine/Skin.java
/******************************************************************************
 * Spine Runtimes License Agreement
 * Last updated April 5, 2025. Replaces all prior versions.
 *
 * Copyright (c) 2013-2025, Esoteric Software LLC
 *
 * Integration of the Spine Runtimes into software or otherwise creating
 * derivative works of the Spine Runtimes is permitted under the terms and
 * conditions of Section 2 of the Spine Editor License Agreement:
 * http://esotericsoftware.com/spine-editor-license
 *
 * Otherwise, it is permitted to integrate the Spine Runtimes into software
 * or otherwise create derivative works of the Spine Runtimes (collectively,
 * "Products"), provided that each user of the Products must obtain their own
 * Spine Editor license and redistribution of the Products in any form must
 * include this license and copyright notice.
 *
 * THE SPINE RUNTIMES ARE PROVIDED BY ESOTERIC SOFTWARE LLC "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL ESOTERIC SOFTWARE LLC BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES,
 * BUSINESS INTERRUPTION, OR LOSS OF USE, DATA, OR PROFITS) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THE SPINE RUNTIMES, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

package com.esotericsoftware.spine;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.Null;
import com.badlogic.gdx.utils.OrderedSet;

import com.esotericsoftware.spine.attachments.Attachment;
import com.esotericsoftware.spine.attachments.MeshAttachment;

/** Stores attachments by slot index and placeholder name. Multiple {@link Skeleton} instances can use the same skins.
 * <p>
 * See {@link SkeletonData#defaultSkin}, {@link Skeleton#skin}, and
 * <a href="https://esotericsoftware.com/spine-runtime-skins">Runtime skins</a> in the Spine Runtimes Guide. */
public class Skin {
	final String name;
	final OrderedSet<SkinEntry> attachments = new OrderedSet();
	final Array<BoneData> bones = new Array(true, 0, BoneData[]::new);
	final Array<ConstraintData> constraints = new Array(true, 0, ConstraintData[]::new);
	private final SkinEntry lookup = new SkinEntry(0, "", null);

	// Nonessential.
	final Color color = new Color(0.99607843f, 0.61960787f, 0.30980393f, 1); // fe9e4fff

	public Skin (String name) {
		if (name == null) throw new IllegalArgumentException("name cannot be null.");
		this.name = name;
		attachments.orderedItems().ordered = false;
	}

	/** Adds an attachment to the skin for the specified slot index and placeholder name. */
	public void setAttachment (int slotIndex, String placeholder, Attachment attachment) {
		if (attachment == null) throw new IllegalArgumentException("attachment cannot be null.");
		var entry = new SkinEntry(slotIndex, placeholder, attachment);
		if (!attachments.add(entry)) attachments.get(entry).attachment = attachment;
	}

	/** Adds all attachments, bones, and constraints from the specified skin to this skin. */
	public void addSkin (Skin skin) {
		if (skin == null) throw new IllegalArgumentException("skin cannot be null.");

		for (BoneData data : skin.bones)
			if (!bones.contains(data, true)) bones.add(data);

		for (ConstraintData data : skin.constraints)
			if (!constraints.contains(data, true)) constraints.add(data);

		for (SkinEntry entry : skin.attachments.orderedItems())
			setAttachment(entry.slotIndex, entry.placeholder, entry.attachment);
	}

	/** Adds all bones and constraints and copies of all attachments from the specified skin to this skin. Mesh attachments are not
	 * copied, instead a new linked mesh is created. The attachment copies can be modified without affecting the originals. */
	public void copySkin (Skin skin) {
		if (skin == null) throw new IllegalArgumentException("skin cannot be null.");

		for (BoneData data : skin.bones)
			if (!bones.contains(data, true)) bones.add(data);

		for (ConstraintData data : skin.constraints)
			if (!constraints.contains(data, true)) constraints.add(data);

		for (SkinEntry entry : skin.attachments.orderedItems()) {
			if (entry.attachment instanceof MeshAttachment mesh)
				setAttachment(entry.slotIndex, entry.placeholder, mesh.newLinkedMesh());
			else
				setAttachment(entry.slotIndex, entry.placeholder, entry.attachment != null ? entry.attachment.copy() : null);
		}
	}

	/** Returns the attachment for the specified slot index and placeholder name, or null. */
	public @Null Attachment getAttachment (int slotIndex, String placeholder) {
		lookup.set(slotIndex, placeholder);
		SkinEntry entry = attachments.get(lookup);
		return entry != null ? entry.attachment : null;
	}

	/** Removes the attachment in the skin for the specified slot index and placeholder name, if any. */
	public void removeAttachment (int slotIndex, String placeholder) {
		lookup.set(slotIndex, placeholder);
		attachments.remove(lookup);
	}

	/** Returns all attachments in this skin. */
	public Array<SkinEntry> getAttachments () {
		return attachments.orderedItems();
	}

	/** Returns all attachments in this skin for the specified slot index. */
	public void getAttachments (int slotIndex, Array<SkinEntry> attachments) {
		if (slotIndex < 0) throw new IllegalArgumentException("slotIndex must be >= 0.");
		if (attachments == null) throw new IllegalArgumentException("attachments cannot be null.");
		for (SkinEntry entry : this.attachments.orderedItems())
			if (entry.slotIndex == slotIndex) attachments.add(entry);
	}

	/** Clears all attachments, bones, and constraints. */
	public void clear () {
		attachments.clear(1024);
		bones.clear();
		constraints.clear();
	}

	public Array<BoneData> getBones () {
		return bones;
	}

	public Array<ConstraintData> getConstraints () {
		return constraints;
	}

	/** The skin's name, unique across all skins in the skeleton.
	 * <p>
	 * See {@link SkeletonData#findSkin(String)}. */
	public String getName () {
		return name;
	}

	/** The color of the skin as it was in Spine, or a default color if nonessential data was not exported. */
	public Color getColor () {
		return color;
	}

	public String toString () {
		return name;
	}

	/** Attach each attachment in this skin if the corresponding attachment in the old skin is currently attached. */
	void attachAll (Skeleton skeleton, Skin oldSkin) {
		Slot[] slots = skeleton.slots.items;
		for (SkinEntry entry : oldSkin.attachments.orderedItems()) {
			SlotPose slot = slots[entry.slotIndex].pose;
			if (slot.attachment == entry.attachment) {
				Attachment attachment = getAttachment(entry.slotIndex, entry.placeholder);
				if (attachment != null) slot.setAttachment(attachment);
			}
		}
	}

	/** Stores an entry in the skin consisting of the slot index and placeholder name. */
	static public class SkinEntry {
		int slotIndex;
		String placeholder;
		@Null Attachment attachment;
		private int hashCode;

		SkinEntry (int slotIndex, String placeholder, @Null Attachment attachment) {
			set(slotIndex, placeholder);
			this.attachment = attachment;
		}

		void set (int slotIndex, String placeholder) {
			if (slotIndex < 0) throw new IllegalArgumentException("slotIndex must be >= 0.");
			if (placeholder == null) throw new IllegalArgumentException("placeholder cannot be null.");
			this.slotIndex = slotIndex;
			this.placeholder = placeholder;
			hashCode = placeholder.hashCode() + slotIndex * 37;
		}

		/** The {@link Skeleton#slots} index. */
		public int getSlotIndex () {
			return slotIndex;
		}

		/** The placeholder name that the attachment is associated with. */
		public String getPlaceholder () {
			return placeholder;
		}

		/** The attachment for this skin entry. */
		public Attachment getAttachment () {
			return attachment;
		}

		public int hashCode () {
			return hashCode;
		}

		public boolean equals (Object object) {
			if (object == null) return false;
			var other = (SkinEntry)object;
			if (slotIndex != other.slotIndex) return false;
			return placeholder.equals(other.placeholder);
		}

		public String toString () {
			return slotIndex + ":" + placeholder;
		}
	}
}
